examples/nlp/text_generation_with_miniature_gpt.py

```diff
@@ -0,0 +1,355 @@
+
+
+def create_model():
+    inputs = layers.Input(shape=(maxlen,), dtype=tf.int32)
+    embedding_layer = TokenAndPositionEmbedding(maxlen, vocab_size, embed_dim)
+    x = embedding_layer(inputs)
+    transformer_block = TransformerBlock(embed_dim, num_heads, feed_forward_dim)
+    x = transformer_block(x)
+    outputs = layers.Dense(vocab_size, activation="softmax")(x)
+    model = keras.Model(inputs=inputs, outputs=[outputs, x])
+    model.compile(
+        "adam", loss=["sparse_categorical_crossentropy", None],  
+    )  # No loss and optimization based on word embeddings from transformer block
+    return model
+
+
+"""
+## Prepare data for word level language modelling
+
+We will download IMDB data, and combine training and validation sets for
+text generation task.
+"""
+
+
+!curl -O https://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz
+!tar -xf aclImdb_v1.tar.gz
+
+batch_size = 16
+
+# The dataset contains each review in a separate text file
+# The text files are present in four different folders
+# Create a list all files
+filenames = []
+directories = ['aclImdb/train/pos', 
+               'aclImdb/train/neg', 
+               'aclImdb/test/pos', 
+               'aclImdb/test/neg']
+for dir in directories:
+    for f in os.listdir(dir):
+        filenames.append(os.path.join(dir, f))
+
+print(f'{len(filenames)} files')
+
+# Create dataset from text files
+text_ds = tf.data.TextLineDataset(filenames)
+text_ds = text_ds.batch(batch_size)
+
+# Create vectcorization layer and adapt it to the text
+vectorize_layer = TextVectorization(
+    standardize=None,
+    max_tokens=vocab_size-1,
+    output_mode="int",
+    output_sequence_length=maxlen+1,
+)
+vectorize_layer.adapt(text_ds)
+vocab = vectorize_layer.get_vocabulary()  # To get words back from token indices
+
+
+def prepare_lm_inputs_labels(text):
+    """
+    Shift word sequences by 1 position so that the target for position (i) is
+    word at position (i+1). The model will use all words up till position (i)
+    to predict the next word.
+    """
+    text = tf.expand_dims(text, -1)
+    tokenized_sentences = vectorize_layer(text)
+    x = tokenized_sentences[:,:-1]
+    y = tokenized_sentences[:,1:]
+    return x, y
+
+text_ds = text_ds.map(prepare_lm_inputs_labels)
+text_ds = text_ds.repeat().prefetch(tf.data.experimental.AUTOTUNE)
+
+
+"""
+## Callback for generating text
+"""
+
+
+class TextGenerator(keras.callbacks.Callback):
+    """
+    Callback to generate text from trained model.
+    1. Feed some starting prompt to the model
+    2. Predict probabilities for next token
+    3. Sample next token and add it to the next input
+    
+    # Arguments
+        max_tokens: Integer, the number of tokens to be generated after prompt.
+        start_tokens: List of integers, the token indices for the starting prompt.
+        index_to_word: List of strings, obatined from TextVectorization layer.
+        temp: Float, dictates the variation during sampling.
```

Comment:
Prefer using "temperature", "temp" is ambiguous since it is usually used for "temporary"